### PR TITLE
Tree dynamics and MST generation code added to Topolotree

### PR DESCRIPTION
**Pull request by Saikrishna-Achalla**
Added some code to topolotree for MST generation and tree dynamics.

- Prims and Kruskals folders contain the respective algorithm implementations to find MST between 24 gcp server regions
- `random_insertions.py` calculates the max node length between left and right subtrees

### CODE REVIEW

File: hydro_cli_examples/examples/topolotree/random_insertions.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+        return 0
+    
+    def helper(root):
+
+        if not root:
+            return 0
+
+        if root.left and root.right:
+            return max(1 + helper(root.left), 1 + helper(root.right))
+        if root.left:
```
Comment (conor-23) [inline hydro_cli_examples/examples/topolotree/random_insertions.py:44]: Its more readable to use "else if" for the cases below "root.left and root.right" .

File: hydro_cli_examples/examples/topolotree/random_insertions.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import random
+import numpy as np
+
```
Comment (conor-23) [inline hydro_cli_examples/examples/topolotree/random_insertions.py:3]: You should include a comment at the top describing what the script does and how we are using it. It will make it easy for people to see if the scripts can be reused or improved on later on in the project.

File: hydro_cli_examples/examples/topolotree/kruskals/default_gcp_regions_mst.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+            else:
+                self.parent[root_y] = root_x
+                self.rank[root_x] += 1
+
+
+def kruskals_algorithm(graph):
+    edges = []
+    for u in range(len(graph)):
+        for v in range(u + 1, len(graph)):
+            if graph[u][v] != 0:
```
Comment (conor-23) [inline hydro_cli_examples/examples/topolotree/kruskals/default_gcp_regions_mst.py:46]: You should add some test cases that you know the correct minimum spanning tree of so that we can run kruskals and prims as well as any future MST algorithms we implement against those.

File: hydro_cli_examples/examples/topolotree/prims/many_locations.py [hunk 4]
```diff
@@ -0,0 +1,62 @@
+import random
+from global_land_mask import globe
+from math import radians, cos, sin, asin, sqrt
+
```
Comment (conor-23) [inline hydro_cli_examples/examples/topolotree/prims/many_locations.py:4]: There is a lot of redundant code across the various py files for prims and kruskals. As a general principle, you don't want to have the same code copy-pasted in two different files. Instead, put it in one place and import that where you need to use it. I would refactor it into a python file containing the code for each of the MST algorithms, a python file containing all the graph helper functions, a test suite of tests to verify the MST algorithms, and a driver program that has the gcp_regions and latencies data in it and launches the html visualizations that you're generating for each algorithm.

**Review by conor-23 — APPROVED** [review]

**Comment by MingweiSamuel** [comment]
@conor-23 Double checking that it makes sense for the topolotree stuff to be here instead of a separate repo?

**Comment by Saikrishna-Achalla** [comment]
Moving to different repo
